dspy/adapters/retry_adapter.py

```diff
@@ -0,0 +1,145 @@
+
+from typing import TYPE_CHECKING, Any, Optional, Type
+import logging
+
+from dspy.adapters.base import Adapter
+from dspy.signatures.signature import Signature
+from dspy.adapters.utils import create_signature_for_retry
+
+if TYPE_CHECKING:
+    from dspy.clients.lm import LM
+
+logger = logging.getLogger(__name__)
+
+class RetryAdapter(Adapter):
+    """
+    RetryAdapter is an adapter that retries the execution of another adapter for
+    a specified number of times if it fails to parse completion outputs.
+    """
+
+    def __init__(self, main_adapter: Adapter, fallback_adapter: Optional[Adapter] = None, max_retries: int = 3):
+        """
+        Initializes the RetryAdapter.
+
+        Args:
+            main_adapter (Adapter): The main adapter to use.
+            fallback_adapter (Optional[Adapter]): The fallback adapter to use if the main adapter fails.
+            max_retries (int): The maximum number of retries. Defaults to 3.
+        """
+        self.main_adapter = main_adapter
+        self.fallback_adapter = fallback_adapter
+        self.max_retries = max_retries
+
+    def __call__(
+        self,
+        lm: "LM",
+        lm_kwargs: dict[str, Any],
+        signature: Type[Signature],
+        demos: list[dict[str, Any]],
+        inputs: dict[str, Any],
+    ) -> list[dict[str, Any]]:
+        """
+        Execute main_adapter and fallback_adapter in the following procedure:
+        1. Call the main_adapter.
+        2. If the main_adapter fails, call the fallback_adapter.
+        3. If the fallback_adapter fails, retry the main_adapter including previous response for `max_retries` times.
+
+        Args:
+            lm (LM): The dspy.LM to use.
+            lm_kwargs (dict[str, Any]): Additional arguments for the lm.
+            signature (Type[Signature]): The signature of the function.
+            demos (list[dict[str, Any]]): A list of demo examples.
+            inputs (dict[str, Any]): A list representating the user input.
+
+        Returns:
+            A list of parsed completions. The size of the list is equal to `n` argument. Defaults to 1.
+
+        Raises:
+            Exception: If fail to parse outputs after the maximum number of retries.
+        """
+        outputs = []
+        max_retries = max(self.max_retries, 0)
+        n_completion = lm_kwargs.get("n", 1)
+
+        values, parse_failures = self._call_adapter(
+            self.main_adapter,
+            lm,
+            lm_kwargs,
+            signature,
+            demos,
+            inputs,
+        )
+        outputs.extend(values)
+
+        if len(outputs) == n_completion:
+            return outputs
+        
+        lm_kwargs["n"] = n_completion - len(outputs)
```

Comment:
I am wondering if it is possible that for the same request's multiple responses, some responses can be parsed while some cannot. It does sound possible, but I don't have this experience... 

Or should we discard all responses if some of responses fail to be parsed? There is a situation where one response has successful parsing because it is incomplete but parseable, e.g., only have `[## Completion ##]` or so.

Reply:
Discussed internally and decided to keep the successful outputs in the previous steps.